jupyter_server_fileid/manager.py

```diff
@@ -122,9 +132,106 @@ def _copy_recursive(self, from_path: str, to_path: str, path_mgr: Any = os.path)
+        if not hasattr(self, "con") or self.con is None:
+            return
+
+        manager_module = self.__class__.__module__
+        manager_classname = self.__class__.__qualname__
+        self.con.execute(
+            "CREATE TABLE Manifest("
+            "manager_module TEXT NOT NULL, "
+            "manager_classname TEXT NOT NULL"
+            ")"
+        )
+        self.con.execute(
+            "INSERT INTO Manifest (manager_module, manager_classname) VALUES (?, ?)",
+            (manager_module, manager_classname),
+        )
+
+    def _check_manifest(self) -> bool:
+        """
+        Returns whether the database file's manifest matches the expected
+        manifest for this class. If one does not exist, writes the manifest and
+        returns True.
+        """
+        if not hasattr(self, "con") or self.con is None:
+            return False
+
+        manifest = None
+        manager_module = self.__class__.__module__
+        manager_classname = self.__class__.__qualname__
+
+        try:
+            manifest = self.con.execute(
+                "SELECT manager_module, manager_classname FROM Manifest"
+            ).fetchone()
+        except Exception:
+            pass
+
+        if manifest is None:
+            self._write_manifest()
+            return True
+
+        manifest_module, manifest_classname = manifest
+        return manifest_module == manager_module and manifest_classname == manager_classname
+
+    def _migrate_database(self) -> None:
+        """Checks whether the database file's manifest matches the expected
+        manifest for this class. Writes the manifest if one does not exist.
+        If one already exists and is incompatible with the expected manifest,
+        then this method backs up the existing database file into a separate
+        file, and then migrates the database over via calling `export_row()` on
+        the old manager and passing the yielded rows to `self.import_rows()`.
+        """
+        if not hasattr(self, "con") or self.con is None:
+            return
+
+        manifest = self.con.execute(
+            "SELECT manager_module, manager_classname FROM Manifest"
+        ).fetchone()
+        prev_module, prev_classname = manifest
+
+        # first, backup the database at backup_db_path.
+        db_path_dir, db_path_basename = os.path.split(self.db_path)
+        backup_db_path = os.path.join(
+            db_path_dir, datetime.datetime.now().strftime("%Y-%m-%dT%H-%M-%S-") + db_path_basename
+        )
+        self.con.close()
+        os.rename(self.db_path, backup_db_path)
+
+        # recreate files table and indices
+        self.con = sqlite3.connect(self.db_path)
+        self._write_manifest()
+        self._create_tables()
+
+        # finally, migrate the database from backup_db_path to the current database.
+        PrevManager = getattr(importlib.import_module(prev_module), prev_classname)
+        for row in PrevManager.export_rows(backup_db_path):
```

Comment:
Seems like `row` should be named _`page`_ or _`batch`_ or something to that effect.  A configurable, defaulting to `1` should also get included, otherwise a code change would be required to increase the page size.

Reply:
We can implement actual batching in a follow-up PR (that also provides benchmarks to justify the new default batch size). For now the batch size is hard-coded to 1.